embedded-graphics/src/primitives/ellipse.rs

```diff
@@ -0,0 +1,518 @@
+        }
+    }
+}
+
+impl Iterator for Points {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ellipse = self.ellipse;
+
+        self.iter.find(|point| ellipse.contains(*point))
+    }
+}
+
+/// Pixel iterator for each pixel in the ellipse border
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    iter: Points,
+
+    // outer_threshold: u32,
+    outer_size: Size,
+    outer_color: Option<C>,
+
+    // inner_threshold: u32,
+    inner_size: Size,
+    inner_color: Option<C>,
+
+    center: Point,
+}
+
+impl<C> StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    fn new(styled: &Styled<Ellipse, PrimitiveStyle<C>>) -> Self {
+        let has_fill_color = styled.style.fill_color.is_some();
+        let has_stroke_color = styled.style.stroke_color.is_some();
+
+        // Always use a stroke width of 0 if no stroke color was set.
+        let stroke_width = if has_stroke_color {
+            styled.style.stroke_width
+        } else {
+            0
+        };
+
+        let outer_size = styled.primitive.size;
+        let inner_size = Size::new(
+            outer_size.width.saturating_sub(2 * stroke_width),
+            outer_size.height.saturating_sub(2 * stroke_width),
+        );
+        let iter = if has_stroke_color || has_fill_color {
+            Points::new(&styled.primitive)
+        } else {
+            Points::empty()
+        };
+
+        Self {
+            iter,
+            outer_size,
+            outer_color: styled.style.stroke_color,
+            inner_size,
+            inner_color: styled.style.fill_color,
+            center: styled.primitive.center_2x(),
+        }
+    }
+}
+
+impl<C> Iterator for StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    type Item = Pixel<C>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self {
+            inner_size,
+            inner_color,
+            outer_color,
+            center,
+            ..
+        } = *self;
+
+        self.iter.find_map(|point| {
+            let inner_threshold = size_to_threshold(&inner_size, &(point * 2 - center));
+
+            let color = if inner_threshold <= 0 {
+                inner_color
+            } else {
+                outer_color
+            };
+
+            color.map(|color| Pixel(point, color))
+        })
+    }
+}
+
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_iter(self)
+    }
+}
+
+/// Uses the ellipse equation b^2 * x^2 + a^2 * y^2 - a^2 * b^2 to return a value signifying whether
+/// a given point lies inside (negative), on (0) or outside (positive) an ellipse centered around
+/// `(0, 0)` with width and height defined by the `size` parameter.
+fn size_to_threshold(size: &Size, point: &Point) -> i32 {
```

Comment:
Is there a reason to use references here? Most other code in e-g doesn't use references for `Size` and `Point`.

Reply:
Thought it might improve performance but benchmarking this compared to passing by copy didn't change much. I've removed the references.